public/app/features/explore/state/datasource.ts

```diff
@@ -0,0 +1,231 @@
+// Libraries
+import { AnyAction, createAction, PayloadAction } from '@reduxjs/toolkit';
+import { RefreshPicker } from '@grafana/ui';
+import { DataSourceApi, HistoryItem } from '@grafana/data';
+import store from 'app/core/store';
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { lastUsedDatasourceKeyForOrgId, stopQueryState } from 'app/core/utils/explore';
+import { ExploreItemState, ThunkResult } from 'app/types';
+
+import { ExploreId } from 'app/types/explore';
+import { getExploreDatasources } from './selectors';
+import { importQueries, runQueries } from './query';
+import { changeRefreshInterval } from './time';
+import { createEmptyQueryResponse, makeInitialUpdateState } from './utils';
+
+/**
+ * Display an error when no datasources have been configured
+ */
+export interface LoadDatasourceMissingPayload {
+  exploreId: ExploreId;
+}
+export const loadDatasourceMissingAction = createAction<LoadDatasourceMissingPayload>('explore/loadDatasourceMissing');
+
+/**
+ * Start the async process of loading a datasource to display a loading indicator
+ */
+export interface LoadDatasourcePendingPayload {
+  exploreId: ExploreId;
+  requestedDatasourceName: string;
+}
+export const loadDatasourcePendingAction = createAction<LoadDatasourcePendingPayload>('explore/loadDatasourcePending');
+
+/**
+ * Datasource loading was completed.
+ */
+export interface LoadDatasourceReadyPayload {
+  exploreId: ExploreId;
+  history: HistoryItem[];
+}
+export const loadDatasourceReadyAction = createAction<LoadDatasourceReadyPayload>('explore/loadDatasourceReady');
+
+/**
+ * Updates datasource instance before datasouce loading has started
+ */
+export interface UpdateDatasourceInstancePayload {
+  exploreId: ExploreId;
+  datasourceInstance: DataSourceApi;
+}
+export const updateDatasourceInstanceAction = createAction<UpdateDatasourceInstancePayload>(
+  'explore/updateDatasourceInstance'
+);
+
+//
+// Action creators
+//
+
```

Comment:
Similarly to this, I would maybe also at the beginning of files add comment "heading" **Actions and Payloads** just to keep it clear and consistent. 
```
//
// Actions and Payloads
//
```